src/main/scala/com/danielasfregola/twitter4s/http/clients/rest/RestClient.scala

```diff
@@ -1,43 +1,102 @@
 package com.danielasfregola.twitter4s.http.clients.rest
 
 import akka.actor.ActorSystem
-import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
+import akka.http.scaladsl.client.RequestBuilding
+import akka.http.scaladsl.model.HttpMethods._
+import akka.http.scaladsl.model._
 import akka.http.scaladsl.unmarshalling.Unmarshal
 import akka.stream.{ActorMaterializer, Materializer}
+import com.danielasfregola.twitter4s.entities.enums.OAuthMode.{MixedAuth, OAuthMode, UseOAuth1, UseOAuthBearerToken}
 import com.danielasfregola.twitter4s.entities.{AccessToken, ConsumerToken, RateLimit, RatedData}
-import com.danielasfregola.twitter4s.http.clients.Client
-import com.danielasfregola.twitter4s.http.oauth.OAuth1Provider
+import com.danielasfregola.twitter4s.http.clients.{BearerTokenClient, Client, OAuthClient}
+import com.danielasfregola.twitter4s.http.marshalling.{BodyEncoder, Parameters}
+import com.danielasfregola.twitter4s.util.Configurations
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
 
-private[twitter4s] class RestClient(val consumerToken: ConsumerToken, val accessToken: AccessToken)(
-    implicit val system: ActorSystem)
-    extends Client {
+private[twitter4s] class RestClient(val authMode: OAuthMode)(implicit val system: ActorSystem)
+    extends Client
+    with RequestBuilding {
+  var v1Client: OAuthClient = _
```

Comment:
`var` should be avoided because they introduce immutability, which is difficult to control in a multi-threaded environment